### PR TITLE
Array broadcasting in arithmetic operations

### LINKED ISSUE #317: Broadcasting in arithmetic operations
## General

* pyfar version: 0.4.2

## Description

During the implementation of PR #277 I stumbled across the following problem:
The docstrings of the arithmetic operations state that combinations of array likes and audio objects are allowed. For N-D signals, this is not quite intuitive. For example, a channelwise amplification raises a broadcasting error
```
a = np.arange(2 * 3 * 4).reshape((2, 3, 4))
b = np.ones((2, 3, 4))
d = a * pf.signals.impulse(10, amplitude=b)
```
So, the internal broadcasting along the time/frequency dimension of the array is missing, while this works:
```
a = np.arange(2 * 3 * 4).reshape((2, 3, 4, 1))
b = np.ones((2, 3, 4))
d = a * pf.signals.impulse(10, amplitude=b)
```
And even this
```
a = np.arange(2 * 3 * 4 * 6).reshape((2, 3, 4, 6))
b = np.ones((2, 3, 4))
d = a * pf.signals.impulse(10, amplitude=b)
```

### PR DESCRIPTION
**Pull request by sikersten**
### Which issue(s) are closed by this pull request?

Closes #317

### Changes proposed in this pull request:

- clarifies arithmetic operations including arrays and audio objects by implementing broadcasting rules
- fixed error for overloaded operators in operations, where the array is first
- add tests for both
- updated docs and concepts

### CODE REVIEW

File: pyfar/_concepts/arithmetic_operations.py [hunk 1]
```diff
@@ -10,22 +10,34 @@
...
+e.g.,
+
+>>> x = np.arange(2 * 3 * 4).reshape((2, 3, 4))
+>>> y = pf.Signal(np.ones((2, 3, 4, 10)), 44100)
+>>> z = pf.add((x, y))
+
+or are broadcasted, e.g.,
+
+>>> x = np.arange(3 * 4).reshape((3, 4))
+>>> y = pf.Signal(np.ones((2, 3, 4, 10)), 44100)
```
Comment (f-brinkmann) [inline pyfar/_concepts/arithmetic_operations.py:39]: To clarify, maybe add a note that y is a signal with `signal.cshape = (2, 3, 4)` and a length of 10 samples.

File: pyfar/classes/audio.py [hunk 2]
```diff
@@ -1197,10 +1238,17 @@ def _get_arithmetic_data(data, n_samples, domain):
...
-
     else:
         data_out = np.asarray(data)
 
+        if data_out.ndim <= len(cshape):
+            # extend by time/frequency axis, scalars are extended to shape (1,)
+            data_out = data_out[..., None]
+        elif cshape != ():
+            # operation includes arrays and audio objects
+            raise ValueError(
```
Comment (f-brinkmann) [inline pyfar/classes/audio.py:1249]: I did not find a test for this - is there one?

**Review by f-brinkmann — APPROVED** [review]
Thanks for detecting and fixing - I only had two small remarks. Otherwise Approved...

File: pyfar/classes/audio.py (hunk 2 above)
Comment (sikersten) [reply pyfar/classes/audio.py:1249]: yes, in ``test_array_broadcasting_errors()``

**Review by mberz — APPROVED** [review]

**Pull request merged**